Refactor/skill settings split

chrisveilleux opened the pull request:
## Description
Complete rethink of how we deal with settings and settingsmeta in core

## How to test
Make changes to settingsmeta.json for a skill.  Confirm that the changes are accurately reflected on the backend.  Make changes to skill settings at account.mycroft.ai.  Confirm that changes are accurately reflected in settings.json

## Contributor license agreement signed?
CLA [yes]

forslund commented:
What's breaking here, The callback registration and the .store() methods? Is there a way of preserving these so we don't break the API without depreciation warning?

The save_settings should only occur if the settings were changed.

Let's move as much of the logic for this outside of the mycroft_skill.

Also let's drop the refactoring commits so it's easier to review.

chrisveilleux commented:
Implemented the change you suggested to make the skill API right again.

I don't think I can move any more of the logic outside of MycroftSkill.  The only thing I really added was a way to deal with settings changes when they occur.

Added logic to only call save_settings if settings change

forslund commented on mycroft/skills/settings.py:
```diff
@@ -12,571 +12,383 @@
+
+
+# TODO: remove in 20.02
+class Settings:
```
Add __getattr__ to keep .get and .update, .keys, etc. methods

```
    def __getattr__(self, attr):
        if attr not in ['store', 'set_changed_callback']:
            return getattr(self._settings, attr)
        else:
            return super().getattr(attr)
```

forslund commented on mycroft/skills/settings.py:
```diff
@@ -12,571 +12,383 @@
-    name = re.sub(r'(^[Ss]kill|[Ss]kill$)', '', name)
-    return camel_case_split(name)
+def save_settings(skill_dir, skill_settings):
+    """Save skill settings to file."""
```
Needs to handle the Settings class specially...

Something along the lines of
```python
def save_settings(skill_dir, skill_settings):                                   
    """Save skill settings to file."""                                          
    settings_path = Path(skill_dir).joinpath('settings.json')                   
                                                                                
    if isinstance(skill_settings, Settings):                                    
        settings = skill_settings.to_dict()                                     
    elif isinstance(skill_settings, dict):                                      
        settings = skill_settings                                               
    else:                                                                       
        raise ValueError('skill_settings should be a dict, not '                
                         '{}'.format(type(skill_settings)))                     
    with open(str(settings_path), 'w') as settings_file:                        
        json.dump(settings, settings_file)
```

chrisveilleux replied:
Why is this necessary?  When I call this method from the settings class, I pass the _settings attribute, which is a dictionary.

forslund replied:
In the future we intend to just pass the MycroftSkill.settings member so we should start doing that directly. Which will show people how to use the function as intended.

Also accessing a _attribute from outside the class should really not be done

forslund commented on mycroft/api/__init__.py:
```diff
@@ -362,49 +358,25 @@ def get_oauth_token(self, dev_cred):
-                        })
-                )
-            return DeviceApi._skill_settings[1]
+    def get_remote_skill_settings(self):
```
I think the "remote" part here is redundant since it's in the DeviceApi it's implied that it's a remote thing.

forslund commented:
Backend has been updated. Merging this now.
